protosocket-server/src/connection_server.rs

```diff
@@ -56,24 +56,120 @@ pub struct ProtosocketServer<Connector: ServerConnector> {
     runtime: tokio::runtime::Handle,
 }
 
+pub struct ProtosocketSocketConfig {
+    nodelay: bool,
+    reuse: bool,
+    keepalive_duration: Option<std::time::Duration>,
+    listen_backlog: u32,
+}
+
+impl ProtosocketSocketConfig {
+    pub fn nodelay(mut self, nodelay: bool) -> Self {
+        self.nodelay = nodelay;
+        self
+    }
+    pub fn reuse(mut self, reuse: bool) -> Self {
+        self.reuse = reuse;
+        self
+    }
+    pub fn keepalive_duration(mut self, keepalive_duration: std::time::Duration) -> Self {
+        self.keepalive_duration = Some(keepalive_duration);
+        self
+    }
+    pub fn listen_backlog(mut self, backlog: u32) -> Self {
+        self.listen_backlog = backlog;
+        self
+    }
+}
+
+impl Default for ProtosocketSocketConfig {
+    fn default() -> Self {
+        Self {
+            nodelay: true,
+            reuse: true,
+            keepalive_duration: None,
+            listen_backlog: 65536,
+        }
+    }
+}
+
+pub struct ProtosocketServerConfig {
+    max_buffer_length: usize,
+    max_queued_outbound_messages: usize,
+    buffer_allocation_increment: usize,
+    socket_config: ProtosocketSocketConfig,
+}
+
+impl ProtosocketServerConfig {
+    pub fn max_buffer_length(mut self, max_buffer_length: usize) -> Self {
+        self.max_buffer_length = max_buffer_length;
+        self
+    }
+    pub fn max_queued_outbound_messages(mut self, max_queued_outbound_messages: usize) -> Self {
+        self.max_queued_outbound_messages = max_queued_outbound_messages;
+        self
+    }
+    pub fn buffer_allocation_increment(mut self, buffer_allocation_increment: usize) -> Self {
+        self.buffer_allocation_increment = buffer_allocation_increment;
+        self
+    }
+    pub fn socket_config(mut self, config: ProtosocketSocketConfig) -> Self {
+        self.socket_config = config;
+        self
+    }
+}
+
+impl Default for ProtosocketServerConfig {
+    fn default() -> Self {
+        Self {
+            max_buffer_length: 16 * (2 << 20),
+            max_queued_outbound_messages: 128,
+            buffer_allocation_increment: 1 << 20,
+            socket_config: Default::default(),
+        }
+    }
+}
+
 impl<Connector: ServerConnector> ProtosocketServer<Connector> {
     /// Construct a new `ProtosocketServer` listening on the provided address.
     /// The address will be bound and listened upon with `SO_REUSEADDR` set.
     /// The server will use the provided runtime to spawn new tcp connections as `protosocket::Connection`s.
     pub async fn new(
-        address: std::net::SocketAddr,
+        address: SocketAddr,
         runtime: tokio::runtime::Handle,
         connector: Connector,
+        config: ProtosocketServerConfig,
```

Comment:
If you're going to break source-compatibility, let's make this api private. Add an `async fn bind_tcp` on ProtosocketServerConfig that delegates to it. This way avoids increasing server initialization complexity.

While we're adding things, let's drop in some documentation on the new structs & functions too!